torch/csrc/api/src/nn/modules/conv.cpp

```diff
@@ -20,119 +25,163 @@ ConvImpl<D, Derived>::ConvImpl(const ConvOptions<D>& options_)
+  }
+  if (!c10::get_if<enumtype::kZeros>(&options.padding_mode())) {
+    stream << ", padding_mode=" << enumtype::get_enum_name(options.padding_mode());
+  }
+  stream << ")";
 }
 
+Conv1dImpl::Conv1dImpl(
+    ConvOptions<1> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
+
 Tensor Conv1dImpl::forward(const Tensor& input) {
-  if (options.transposed()) {
-    return torch::conv_transpose1d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.output_padding(),
-        options.groups(),
-        options.dilation());
+  if (c10::get_if<enumtype::kCircular>(&options.padding_mode())) {
+    std::vector<int64_t> expanded_padding = {((*options.padding())[0] + 1) / 2, (*options.padding())[0] / 2};
+    return F::conv1d(
+      F::pad(input, PadOptions(expanded_padding).mode(torch::kCircular)),
+      weight, bias,
+      Conv1dOptions()
+        .stride(options.stride())
+        .padding(0)
+        .dilation(options.dilation())
+        .groups(options.groups()));
   }
-  return torch::conv1d(
-      input,
-      weight,
-      bias,
-      options.stride(),
-      options.padding(),
-      options.dilation(),
-      options.groups());
+  return F::conv1d(
+    input,
+    weight,
+    bias,
+    Conv1dOptions()
+      .stride(options.stride())
+      .padding(options.padding())
+      .dilation(options.dilation())
+      .groups(options.groups()));
 }
 
+Conv2dImpl::Conv2dImpl(
+    ConvOptions<2> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
+
 Tensor Conv2dImpl::forward(const Tensor& input) {
-  if (options.transposed()) {
-    return torch::conv_transpose2d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.output_padding(),
-        options.groups(),
-        options.dilation());
+  if (c10::get_if<enumtype::kCircular>(&options.padding_mode())) {
+    std::vector<int64_t> expanded_padding = {
+      ((*options.padding())[1] + 1) / 2, (*options.padding())[1] / 2,
+      ((*options.padding())[0] + 1) / 2, (*options.padding())[0] / 2};
+    return F::conv2d(
+      F::pad(input, PadOptions(expanded_padding).mode(torch::kCircular)),
+      weight, bias,
+      Conv2dOptions()
+        .stride(options.stride())
+        .padding(0)
+        .dilation(options.dilation())
+        .groups(options.groups()));
   }
-  return torch::conv2d(
-      input,
-      weight,
-      bias,
-      options.stride(),
-      options.padding(),
-      options.dilation(),
-      options.groups());
+  return F::conv2d(
+    input,
+    weight,
+    bias,
+    Conv2dOptions()
+      .stride(options.stride())
+      .padding(options.padding())
+      .dilation(options.dilation())
+      .groups(options.groups()));
 }
 
+Conv3dImpl::Conv3dImpl(
+    ConvOptions<3> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
```

Comment:
It is because we pass `options_` by value here, as a result we can use `std::move` as an optimization. The reason for passing `options_` by value is that we want to be able to change `options_` in-place (updating its `transposed_` and `output_padding_` fields) before passing it to the `ConvImpl` constructor.